---
title: "GLSconversion"
author: "Derek Pappas, Ph.D. (dpappas@chori.org)"
date: "August 21, 2017"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Using GLSconversion}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

## Overview

Tools for cross converting HLA or KIR genotyping data from text strings to multi-column table format as desribed in Milius RP, Mack SJ, Hollenbach JA, et al. 2013. Genotype List String: a grammar for describing HLA and KIR genotyping results in a text string. [Tissue Antigens. 82:106-112](http://www.ncbi.nlm.nih.gov/pubmed/23849068). 

### Input Data

Data may be passed to GLSconvert as a tab delimited text file name (full path recommended) or as a R object (data frame). Whether a text file or an R object, the first row must be a header line and include column names for the GL string/locus genotypes. The columns of miscellanous information preceeding the genotype column(s) may contain any information, but generally includes at least the sample id. While there is no limit to the number of columns, there is a requirement to column ordering depending on the conversion direction.  

_**GL String to Table**_  
Formatting for GL string conversion requires that the last column of the data table must contain the GL string.

|SubjectID |Exp ID  |GLString                                       |
|----------|:------:|:---------------------------------------------:|
|Subject1  |Center1 |HLA-A\*01:01+HLA-A\*02:01\^HLA-B\*08:01+HLA-B\*44:02\^HLA-DRB1\*01:01+HLA-DRB1\*03:01 |


_**Table to GL String**_  
Formatting for table conversion requires three columns. At least one column of identifying information and subsequent column pairs for each locus. Genotype locus pairs must be located in adjacent columns. Column names for a given locus may use (not required) '_1', '.1','_2','.2' to distinguish each locus pair. Only column names for each locus may repeat, all other column names must be unique. You may format your alleles as Locus*Allele or Allele following defined HLA and KIR naming conventions. 

|SubjectID |Exp ID  |   A   |   A   |   B   |   B   | DRB1  | DRB1  | DRB3  | DRB3  |
|----------|:------:|:-----:|:-----:|:-----:|:-----:|:-----:|:-----:|:-----:|:-----:|
|Subject1  |Center1 |01:01  |02:01  |08:01  |44:02  |01:01  |04:01  |       |       |


**Homozygosity**  
Homozygous allele calls for non-DRB345 alleles can be represented as single alleles in the GL string. For example, HLA-A\*01:01:01:01 + HLA-A\*01:01:01:01 can be written as HLA-A*01:01:01:01.  

**HLA-DRB3, HLA-DRB4, and HLA-DRB5**  
HLA-DRB3, HLA-DRB4, and HLA-DRB5 are parsed for homozygous or hemizygous status based on the DRB1 haplotype as defined by Andersson, 1998 (Andersson G. 1998. Evolution of the HLA-DR region. [Front Biosci. 3:d739-45](http://www.ncbi.nlm.nih.gov/pubmed/9675159)) and can be flagged for inconsistency. Inconsistent haplotypes are identified with an "!" appened to the allele (in table format) or the GL string (in GL string format).

**Ambiguity**
The GL2Tab will convert GL strings with genotype amibguity ("|") to mutlitple lines within the final table to be output. The Tab2GL conversion requires amibiguity to exists as multiple (2 or more) rows within the table using identical miscellaneous information (columns preceeding genotyping data). If the data within these columns are not identical, they will not be flagged for inclusion in an ambiguity string.

|SubjectID |Exp ID  |   A   |   A   |   B   |   B   | DRB1  | DRB1  | DPB1  | DPB1  |
|----------|:------:|:-----:|:-----:|:-----:|:-----:|:-----:|:-----:|:-----:|:-----:|
|Subject1  |Center1 |01:01  |02:01  |08:01  |44:02  |01:01  |04:01  |04:02:01:01|02:01:02|
|Subject1  |Center1 |01:01  |02:01  |08:01  |44:02  |01:01  |04:01  |105:01|416:01|


## Data Output

Data can be output to either a text file (tab or comma delimitted) or R object (sent to data frame). See Output parameters below. When running the GL2Tab conversion, all adjacent pairs of loci will include '_1' and '_2' to distinguish each chromosome. Please not, subsequent programs used to analyze the data table such as BIGDAWG or Pypop may not accept files with ambiguous genotyping data.


## Parameters

`GLS.Convert(Data,Convert,Output="txt",System="HLA",HZY.Red=FALSE,DRB345.Flag=FALSE,Cores.Lim=1L)`

**Data**

Class: String/Object. (No Default).

e.g., Data="/your/path/to/file/foo.txt" -or- Data="foo.txt" -or- Data=foo (No Default)

Specifies data file name or data object. File name is either full file name path to specify file location (recommended) or name of file within a set working directory. See Data Input section for details about file formatting. **This parameter is required for the conversion utility.**  

**Convert**

Class: String. Options: "GL2Tab" -or- "Tab2GL" (No Default).

Specifies data file name or data object. May use file name within working directory or full file name path to specify file location (recommended). See Data Input section for details about file formatting. **This parameter is required for the conversion utility.**  

**Output**

Class: String. Options: "R" -or- "txt" -or- "csv" -or- "pypop" (Default = "txt").

Specifies the type of output for the converted genotypes. For file writting, if you specified the full path for a file name then the resultant file will be written to the same directory. Otherwise the file will be written to whichever working directory was defined at initiation of conversion. The converted file name will be of the form "Converted_foo.txt" depending on output setting. If the data was an R object, the file name will be "Converted.txt" if output to file is desired. To output as R object will require an assignment to some object (see examples below).

**System**

Class: String. Options: "HLA" or "KIR" (Default="HLA").

Defines the genetic system of the data being converted. This parameter is required for Tab2GL conversion and is ignored for GL2Tab. The default system is HLA.

**HZY.Red**

Class: Logical (Default=FALSE).

Homozygous reduction: Should non-dRB345 homozygotes be represent by a single allele name in GL string? For example: HLA-A*01:01:01:01+HLA-A*01:01:01:01 as HLA-A*01:01:01:01. The default behavior to keep both allele names in the GL string. This setting does not impact DRB3, DRB4, or DRB5 genotype calls nor will it impact the genotype list to table conversion (GL2Tab).  

**DRB345.Flag**

Class: Logical (Default=FALSE).

Indicates whether DR haplotypes should be parsed for correct zygosity and unusual DR haplotypes flagged with an exclamation ('!') appended to genotype. The default behaviour will not flag unusual haplotypes.

## Examples

These are examples only and need not be run as defined below.

```
# Installation of GLSconversion from GitHub
# May require 'devtools' package installation first.
# May require closing and reopening of R Studio after install.
install.packages("devtools") # optional
library("devtools")
install_github("IgDAWG/GLSconversion", build_vignettes = TRUE) # Requires Pandoc or RStudio
install_github("IgDAWG/GLSconversion") # No Pandoc or RStudio

# Run the a GL2Tab conversion on a data file with default output to text file
GLSconvert(Data="/your/path/to/file/foo.txt", conv="GL2Tab")

# Run the a Tab2GL conversion on a R object outputting to a R object
foo.tab <- GLSconvert(Data=foo, conv="Tab2GL")

# Run the a Tab2GL conversion on a text file outputting to a csv file
foo.tab <- GLSconvert(Data=foo, conv="Tab2GL", output="csv")

# Run the a GL2Tab conversion on a data file without the full path name and with homozygous allele reduction
setwd("/your/path/to/file")
GLSconvert(Data="foo.txt", conv="GL2Tab", HZY.Red=TRUE)

```

*End of vignette.*

